File: 268 Missing Number/268.py
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        # lowcase set not Set !
        hash_set = set(nums)
        for i in range(len(nums) + 1):
            if i not in hash_set:
                return i
        return -1
    
    def missingNumber2(self, nums: List[int]) -> int:
        miss = len(nums)
        for i, num in enumerate(nums):
            miss ^= i ^ num
        return miss

File: 268 Missing Number/268.java
class Solution {
    // Solution 1: Hash Set
    public int missingNumber(int[] nums) {
        Set<Integer> hashSet = new HashSet<>();
        for (int num : nums) {
            hashSet.add(num);
        }
        for (int i = 0; i <= nums.length; ++i) {
            if (!hashSet.contains(i)) {
                return i;
            }
        }
        return -1;
    }

    // Time Complexity: O(2n) -> O(n) in average, worse case O(n^2)
    // Space Complexity: O(n)

    // Solution 2: Bit XOR
    public int missingNumber2(int[] nums) {
        int miss = nums.length;
        for (int i = 0; i < nums.length; ++i) {
            miss ^= nums[i] ^ i;
        }
        return miss;
    }

    // Time Complexity: O(n)
    // Space Complexity: O(1)
}